docs/RFCS/online_schema_change.md

```diff
@@ -0,0 +1,144 @@
+are in use concurrently. The table lease mechanism will be used to
+ensure that there are only two active versions of a descriptor in use
+in a cluster and the scheme change implementation will ensure that
+these two versions follow the state transition invariant.
+
+When an index is added to a table a potentially time consuming
+backfill operation needs to be performed to create the index
+data. This operation is performed when the index is in the
+`WRITE_ONLY` state and completed before moving the index to the
+`PUBLIC` state. Backfilling creates index entries for every row in the
+table. As such, it is infeasible for this backfilling to be performed
+transactionally: the transaction would create an unreasonable number
+of write intents and user transactions could cause the transaction to
+abort. Instead, backfill be performed as a series of small
+transactions that process a fraction of the table at a time:
+
+```go
+  startKey := MakeTableIndexKey(desc.ID, desc.Primary)
+  endKey := startKey.PrefixEnd()
+  for startKey != endKey {
+    var lastKey roachpb.Key
+    err := db.Txn(func(txn *Txn) error {
+      txn.SetPriority(VeryLowPriority)
+      scan, err := txn.Scan(startKey, endKey, 1000)
+      if err != nil {
+        return err
+      }
+      lastKey = getLastKeyOfFullRowInScan(scan)
+      b := makeIndexKeysBatch(scan)
+      return txn.CommitInBatch(b)
+    })
+    if err != nil {
+      // Abort!
+    }
+    startKey = lastKey.Next()
+  }
+```
+
+The above pseudo-code is intended to give the gist of how backfilling
+will work. We'll iterate over the primary indexn transactionally
+retrieving keys and generating index entries. We perform the
+backfilling work transactionally in order to avoid anomalies involving
+deletions. Backfilling might duplicate work performed by concurrent
+insertions or updates, but the blind overwrites of identical data are
+safe.
+
+Since schema change operations are potentially long running they need
+to be restartable or abortable if the node performing them dies. We
+accomplish this by performing the schema change operation for a table
+on a well known node: the replica holding the leader lease for the
+first range of the table (i.e. containing the key `/<tableID>`). When
+a node receives a schema change operation such as `CREATE INDEX` it
+will forward the operation to this "table leader". When the table
+leader restarts it will load the associated table descriptor and
+restart or abort the schema change operation.
+
+For initial implementation simplicity, we will only allow a single
+schema change operation at a time per table. This restriction can be
+lifted in the future if we see a benefit in allowing concurrent schema
+changes on a table (e.g. concurrently adding multiple indexes).
+
+# Drawbacks
+
+# Alternatives
+
+# Unresolved questions
+
+* Is there a way to avoid performing the backfilling work
+  transactionally? Is it worth optimizing?
```

Comment:
A more obvious optimization is persisting which key ranges have already been backfilled, so that an interrupted backfill avoids duplicate work.